libcli.c

```diff
@@ -1966,11 +1990,1146 @@ void cli_set_idle_timeout_callback(struct cli_def *cli, unsigned int seconds, in
+    free(c->command);
+    free(c);
+    return NULL;
+  }
+  // buildmode commmands are all registered at the top level
+  if (!cli_register_command_core(cli, NULL, c)) {
+    cli_free_command(c);
+    c = NULL;
+  }
+  return c;
+}
+
+int cli_int_execute_buildmode(struct cli_def *cli) {
+  struct cli_optarg *optarg = NULL;
+  int rc = CLI_OK;
+  char *cmdline;
+
+  char *value = NULL;
+
+  cmdline = strdup(cli_command_name(cli, cli->buildmode->command));
+  for (optarg = cli->buildmode->command->optargs; rc == CLI_OK && optarg; optarg = optarg->next) {
+    value = NULL;
+    do {
+      if (cli->privilege < optarg->privilege) continue;
+      if ((optarg->mode != cli->buildmode->mode) && (optarg->mode != cli->buildmode->transient_mode) &&
+          (optarg->mode != MODE_ANY))
+        continue;
+
+      value = cli_get_optarg_value(cli, optarg->name, value);
+      if (!value && optarg->flags & CLI_CMD_ARGUMENT) {
+        cli_error(cli, "Missing required argument %s", optarg->name);
+        rc = CLI_MISSING_ARGUMENT;
+      } else if (value) {
+        if (optarg->flags & (CLI_CMD_OPTIONAL_FLAG | CLI_CMD_ARGUMENT)) {
+          if (!(cmdline = cli_int_buildmode_extend_cmdline(cmdline, value))) {
+            cli_error(cli, "Unable to append to building commandlne");
+            rc = CLI_ERROR;
+          }
+        } else {
+          if (!(cmdline = cli_int_buildmode_extend_cmdline(cmdline, optarg->name))) {
+            cli_error(cli, "Unable to append to building commandlne");
+            rc = CLI_ERROR;
+          }
+          if (!(cmdline = cli_int_buildmode_extend_cmdline(cmdline, value))) {
+            cli_error(cli, "Unable to append to building commandlne");
+            rc = CLI_ERROR;
+          }
+        }
+      }
+    } while (rc == CLI_OK && value && optarg->flags & CLI_CMD_OPTION_MULTIPLE);
+  }
+
+  if (rc == CLI_OK) {
+    cli_int_free_buildmode(cli);
+    cli_add_history(cli, cmdline);
+    rc = cli_run_command(cli, cmdline);
+  }
+  free_z(cmdline);
+  cli_int_free_buildmode(cli);
+  return rc;
+}
+
+char *cli_int_buildmode_extend_cmdline(char *cmdline, char *word) {
+  char *tptr = NULL;
+  char *cptr = NULL;
+  size_t oldlen = strlen(cmdline);
+  size_t wordlen = strlen(word);
+  int add_quotes = 0;
+
+  /*
+   * Allocate enough space to hold the old string, a space, and the new string (including null terminator).
+   * Also include enough space for a quote around the string if it contains a whitespace character
+   */
+  if ((tptr = (char *)realloc(cmdline, oldlen + 1 + wordlen + 1 + 2))) {
+    strcat(tptr, " ");
+    for (cptr = word; *cptr; cptr++) {
+      if (isspace(*cptr)) {
+        add_quotes = 1;
+        break;
+      }
+    }
+    if (add_quotes) strcat(tptr, "'");
+    strcat(tptr, word);
+    if (add_quotes) strcat(tptr, "'");
+  }
+  return tptr;
+}
+
+int cli_int_buildmode_cmd_cback(struct cli_def *cli, const char *command, char *argv[], int argc) {
+  int rc = CLI_BUILDMODE_EXTEND;
```

Comment:
Here and below in a few places. This file is already too long!

```c
 if (argc) {
    cli_error(cli, "Extra arguments on command line, command ignored.");
    return CLI_ERROR;
  }
  return CLI_BUILDMODE_EXTEND;
```

Reply:
Concur that libcli.c is a long file, but having everything in a monolithic compilation unit helps with not exposing things we don't want to.  But I don't understand the 'here and below...' comment.